pkg/aggregator/internal/limiter/limiter.go

```diff
@@ -0,0 +1,165 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package limiter
+
+import (
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/tagset"
+)
+
+type entry struct {
+	current  int // number of contexts currently in aggregator
+	rejected int // number of rejected samples
+	tags     []string
+}
+
+// Limiter tracks number of contexts based on origin detection metrics
+// and rejects samples if the number goes over the limit.
+//
+// Not thread safe.
+type Limiter struct {
+	key   string
+	tags  []string
+	limit int
+	usage map[string]*entry
+}
+
+// New returns a new instance of limiter.
+//
+// If limit is zero or less the limiter is disabled.
+func New(limit int, key string, tags []string) *Limiter {
+	if limit <= 0 {
+		return nil
+	}
+
+	if !strings.HasSuffix(key, ":") {
+		key += ":"
+	}
+
+	hasKey := false
+	tags = append([]string{}, tags...)
+	for i := range tags {
+		if !strings.HasSuffix(tags[i], ":") {
+			tags[i] += ":"
+		}
+		hasKey = hasKey || key == tags[i]
+	}
+
+	if !hasKey {
+		tags = append(tags, key)
+	}
+
+	return &Limiter{
+		key:   key,
+		tags:  tags,
+		limit: limit,
+		usage: map[string]*entry{},
+	}
+}
+
+func (l *Limiter) identify(tags []string) string {
+	for _, t := range tags {
+		if strings.HasPrefix(t, l.key) {
+			return t
+		}
+	}
+	return ""
+}
+
+func (l *Limiter) extractTags(src []string) []string {
+	dst := make([]string, 0, len(l.tags))
```

Comment:
Here's what I understand this function to be doing:
- The set of `l.tags` are tag keys that we'd like to report in telemetry for dropped contexts
- Given a context (ie, set of tags), we'd like to save the tag values that appear in the `l.tags`
- So find all `src` tags that have a prefix that appears in `l.tags` and return those.